Magiczne-Schody/UT_ObserverPattern.cpp

```diff
@@ -0,0 +1,165 @@
+
+#include "stdafx.h"
+#include "CppUnitTest.h"
+#include "Observer.h"
+#include "LightUp.h"
+#include "LightDown.h"
+#include "TurnOff.h"
+
+using namespace Microsoft::VisualStudio::CppUnitTestFramework;
+
+namespace ObserverPatternTest
+{
+	bool testVar = 0;
+	void TestCallback(void) {
+		testVar = !testVar;
+	}
+
+	TEST_CLASS(ObserverPattern_TEST)
+	{
+	public:
+		TEST_METHOD(Subscribing)
+		{
+			int numberOfObservers = 0;
+			LightUp Observer1;
+			LightDown Observer2;
+			TurnOff Observer3;
+			Observable Notifier;
+
+			Observer1.SubscribeToObservable(&Notifier);
+			Observer2.SubscribeToObservable(&Notifier);
+			Observer3.SubscribeToObservable(&Notifier);
+
+			numberOfObservers = Notifier.CheckHowManyObserversRegistered();
+			Assert::AreEqual(3, numberOfObservers);
+		}
+
+		TEST_METHOD(NotSubscribingIfAlreadySubsribed)
+		{
+			int numberOfObservers = 0;
+			LightUp Observer1;
+			LightDown Observer2;
+			TurnOff Observer3;
+			Observable Notifier;
+
+			Observer1.SubscribeToObservable(&Notifier);
+			Observer2.SubscribeToObservable(&Notifier);
+			Observer3.SubscribeToObservable(&Notifier);
+			Observer3.SubscribeToObservable(&Notifier);
+			Observer1.SubscribeToObservable(&Notifier);
+
+			numberOfObservers = Notifier.CheckHowManyObserversRegistered();
+			Assert::AreEqual(3, numberOfObservers);
+		}
+
+
+		TEST_METHOD(Unsubscribing)
+		{
+			int numberOfObservers = 0;
+			LightDown Observer1;
+			LightUp Observer2;
+			TurnOff Observer3;
+			Observable Notifier;
+
+			Observer1.SubscribeToObservable(&Notifier);
+			Observer2.SubscribeToObservable(&Notifier);
+			Observer3.SubscribeToObservable(&Notifier);
+
+			Observer1.Unsubscribe(&Notifier);
+			Observer3.Unsubscribe(&Notifier);
+
+			numberOfObservers = Notifier.CheckHowManyObserversRegistered();
+			Assert::AreEqual(1, numberOfObservers);
+
+			Notifier.NotifyObservers();
+
+			Assert::AreEqual(static_cast<float>(100), Observer::arrayWithBrightnessValues[0]);
+		}
+
+
+		TEST_METHOD(UnregisterAll)
+		{
+			int numberOfObservers = 0;
+			TurnOff Observer1;
+			LightDown Observer2;
+			LightUp Observer3;
+			Observable Notifier;
+
```

Comment:
`Observer::arrayWithBrightnessValues[0] = 0;` - to rozwiąże problem z wynikiem testu, który czasami jest pozytywny a czasami negatywny w zależności czy uruchomimy tylko ten test, czy jest on poprzedzony innymi. Prawdopodobnie dzieje się tak, ponieważ w pliku Obserrver.cpp inicjalizujesz tablicę ale dzięki temu, że jest to składnik statyczny jest tworzony tylko raz (dla wszystkich unit testów) i w dodatku jest tworzony zanim powstanie jakikolwiek obiekt klasy Observer.  Jeżeli chcemy wyzerować tablicę (zgasić wszystkie światła) powinniśmy to zrobić za pomocą jakiegoś efektu wygaszającego. Albo po prostu ręcznie tak jak to powinno być zrobione w tym teście :)

Reply:
:)
Chyba nie puszczałam wszystkich testów na raz i dlatego tego nie zauważyłam... Dzięki!